File: app/src/main/java/com/azreashade/shadowpurge/ExclusionManager.java
package com.azreashade.shadowpurge;

import android.content.Context;
import android.content.SharedPreferences;
import java.util.HashSet;
import java.util.Set;

public class ExclusionManager {
    private static final String PREF = "shadow_prefs";
    private static final String KEY = "excluded_pkgs";
    private final SharedPreferences prefs;

    public ExclusionManager(Context ctx) {
        prefs = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
    }

    public Set<String> getExcluded() {
        return new HashSet<>(prefs.getStringSet(KEY, new HashSet<String>()));
    }

    public void setExcluded(Set<String> pkgs) {
        prefs.edit().putStringSet(KEY, pkgs).apply();
    }

    public boolean isExcluded(String pkg) {
        return getExcluded().contains(pkg);
    }

    public void toggle(String pkg) {
        Set<String> s = getExcluded();
        if (s.contains(pkg)) s.remove(pkg); else s.add(pkg);
        setExcluded(s);
    }
}

File: app/src/main/java/com/azreashade/shadowpurge/data/ExclusionManager.kt
package com.azreashade.shadowpurge.data

import android.content.Context
import android.content.SharedPreferences

class ExclusionManager(context: Context) {

    companion object {
        private const val PREFS_NAME = "shadow_purge_prefs"
        private const val KEY_EXCLUDED_APPS = "excluded_apps"
    }

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private var excludedApps: MutableSet<String> = loadExcludedApps()

    private fun loadExcludedApps(): MutableSet<String> {
        return prefs.getStringSet(KEY_EXCLUDED_APPS, mutableSetOf())?.toMutableSet() ?: mutableSetOf()
    }

    private fun saveExcludedApps() {
        prefs.edit().putStringSet(KEY_EXCLUDED_APPS, excludedApps).apply()
    }

    fun isExcluded(packageName: String): Boolean {
        return excludedApps.contains(packageName)
    }

    fun toggleExclusion(packageName: String) {
        if (excludedApps.contains(packageName)) {
            excludedApps.remove(packageName)
        } else {
            excludedApps.add(packageName)
        }
        saveExcludedApps()
    }

    fun getAllExcluded(): Set<String> = excludedApps.toSet()

    fun setExcludedApps(newSet: Set<String>) {
        excludedApps = newSet.toMutableSet()
        saveExcludedApps()
    }
}
